File: app/src/main/java/com/example/purepawapp/ui/order/OrderDetailFragment.java
package com.example.purepawapp.ui.order;

import android.graphics.Typeface;
import android.os.Bundle;
import android.view.Gravity;
import android.view.LayoutInflater;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.TextView;

import androidx.navigation.fragment.NavHostFragment;

import com.example.purepawapp.R;
import com.example.purepawapp.data.model.Order;
import com.example.purepawapp.data.model.OrderItem;
import com.example.purepawapp.data.model.OrderStatusEvent;
import com.example.purepawapp.data.repository.RepoCallback;
import com.example.purepawapp.databinding.FragmentOrderDetailBinding;
import com.example.purepawapp.databinding.ItemAdminOrderLineBinding;
import com.example.purepawapp.di.ServiceLocator;
import com.example.purepawapp.ui.admin.AdminStatusUi;
import com.example.purepawapp.ui.admin.StatusStyle;
import com.example.purepawapp.ui.common.BaseFragment;
import com.example.purepawapp.util.CurrencyUtils;
import com.example.purepawapp.util.ViewUtils;

import java.text.SimpleDateFormat;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.List;
import java.util.Locale;

public class OrderDetailFragment extends BaseFragment<FragmentOrderDetailBinding> {

    private String orderId;
    private final SimpleDateFormat dateFormat = new SimpleDateFormat("dd/MM/yyyy", new Locale("vi", "VN"));
    private final SimpleDateFormat dateTimeFormat = new SimpleDateFormat("HH:mm, dd/MM/yyyy", new Locale("vi", "VN"));

    public OrderDetailFragment() {
        super(FragmentOrderDetailBinding::inflate);
    }

    @Override
    public void onViewCreated(View view, Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        orderId = getArguments() != null ? getArguments().getString("orderId", "") : "";

        getBinding().btnBack.setOnClickListener(v -> NavHostFragment.findNavController(this).popBackStack());

        loadOrder();
    }

    @Override
    public void onResume() {
        super.onResume();
        loadOrder();
    }

    private void loadOrder() {
        if (orderId == null || orderId.isBlank()) return;
        showLoading();
        ServiceLocator.getOrderRepository().getOrder(orderId, new RepoCallback<>() {
            @Override
            public void onSuccess(Order order) {
                bind(order);
                hideLoading();
            }

            @Override
            public void onError(Exception error) {
                ViewUtils.toast(OrderDetailFragment.this, error.getMessage() != null ? error.getMessage() : "Không thể tải đơn hàng");
                hideLoading();
            }
        });
    }

    private void bind(Order order) {
        StatusStyle style = AdminStatusUi.orderStatusStyle(order.getStatus());
        getBinding().tvHeroTitle.setText(style.getLabel());
        getBinding().tvHeroSubtitle.setText("Ngày đặt: " + dateFormat.format(order.getCreatedAt()));
        String orderCode = order.getOrderCode();
        String displayCode = orderCode != null && !orderCode.isBlank() ? orderCode : "#" + order.getId().substring(0, Math.min(8, order.getId().length()));
        getBinding().tvOrderCode.setText("Mã đơn: " + displayCode);

        renderTimeline(order);

        getBinding().llOrderItems.removeAllViews();
        for (OrderItem item : order.getItems()) {
            ItemAdminOrderLineBinding lineBinding = ItemAdminOrderLineBinding.inflate(LayoutInflater.from(requireContext()), getBinding().llOrderItems, false);
            lineBinding.tvLineName.setText(item.getProductName() + " x" + item.getQuantity());
            lineBinding.tvLineSubtotal.setText(CurrencyUtils.toVndString(item.getSubtotal()));
            getBinding().llOrderItems.addView(lineBinding.getRoot());
        }

        var address = order.getShippingAddress();
        getBinding().tvShippingName.setText(address.getFullName() + " | " + address.getPhone());
        List<String> addressParts = new ArrayList<>();
        for (String part : List.of(address.getStreet(), address.getWard(), address.getDistrict(), address.getCity())) {
            if (part != null && !part.isBlank()) addressParts.add(part);
        }
        getBinding().tvShippingAddress.setText(String.join(", ", addressParts));

        String paymentMethod = order.getPaymentMethod() == null ? "" : order.getPaymentMethod();
        String paymentLabel;
        switch (paymentMethod.toLowerCase()) {
            case "cod":
                paymentLabel = "Thanh toán khi nhận hàng (COD)";
                break;
            case "bank":
            case "banking":
                paymentLabel = "Chuyển khoản ngân hàng";
                break;
            default:
                paymentLabel = paymentMethod.isBlank() ? "Thanh toán khi nhận hàng (COD)" : paymentMethod;
        }
        getBinding().tvPaymentMethod.setText(paymentLabel);

        getBinding().tvSubtotal.setText(CurrencyUtils.toVndString(order.getSubtotal()));
        getBinding().tvShippingFee.setText(CurrencyUtils.toVndString(order.getShippingFee()));
        if (order.getDiscountAmount() > 0) {
            getBinding().rowDiscount.setVisibility(View.VISIBLE);
            getBinding().tvDiscount.setText("-" + CurrencyUtils.toVndString(order.getDiscountAmount()));
        } else {
            getBinding().rowDiscount.setVisibility(View.GONE);
        }
        getBinding().tvTotalAmount.setText(CurrencyUtils.toVndString(order.getTotalAmount()));
    }

    private void renderTimeline(Order order) {
        getBinding().llTimeline.removeAllViews();
        float density = getResources().getDisplayMetrics().density;

        List<OrderStatusEvent> history = new ArrayList<>(order.getStatusHistory());
        if (history.isEmpty()) {
            history.add(new OrderStatusEvent(order.getStatus(), order.getCreatedAt()));
        }
        history.sort(Comparator.comparingLong(OrderStatusEvent::getTimestamp).reversed());

        for (int index = 0; index < history.size(); index++) {
            OrderStatusEvent event = history.get(index);
            boolean isLast = index == history.size() - 1;

            LinearLayout row = new LinearLayout(requireContext());
            row.setOrientation(LinearLayout.HORIZONTAL);
            row.setLayoutParams(new LinearLayout.LayoutParams(LinearLayout.LayoutParams.MATCH_PARENT, LinearLayout.LayoutParams.WRAP_CONTENT));

            LinearLayout dotColumn = new LinearLayout(requireContext());
            dotColumn.setOrientation(LinearLayout.VERTICAL);
            dotColumn.setGravity(Gravity.CENTER_HORIZONTAL);
            dotColumn.setLayoutParams(new LinearLayout.LayoutParams(
                    (int) (12 * density),
                    isLast ? LinearLayout.LayoutParams.WRAP_CONTENT : LinearLayout.LayoutParams.MATCH_PARENT));

            View dot = new View(requireContext());
            dot.setBackground(getResources().getDrawable(R.drawable.shape_circle, null));
            dot.setBackgroundTintList(getResources().getColorStateList(
                    index == 0 ? R.color.pp_primary : R.color.pp_timeline_inactive, null));
            dot.setLayoutParams(new LinearLayout.LayoutParams((int) (12 * density), (int) (12 * density)));
            dotColumn.addView(dot);

            if (!isLast) {
                View line = new View(requireContext());
                line.setBackgroundColor(getResources().getColor(R.color.pp_timeline_inactive, null));
                LinearLayout.LayoutParams lineParams = new LinearLayout.LayoutParams((int) (2 * density), 0, 1f);
                line.setLayoutParams(lineParams);
                dotColumn.addView(line);
            }

            LinearLayout textColumn = new LinearLayout(requireContext());
            textColumn.setOrientation(LinearLayout.VERTICAL);
            LinearLayout.LayoutParams textParams = new LinearLayout.LayoutParams(0, LinearLayout.LayoutParams.WRAP_CONTENT, 1f);
            textParams.leftMargin = (int) (16 * density);
            if (!isLast) textParams.bottomMargin = (int) (20 * density);
            textColumn.setLayoutParams(textParams);

            StatusStyle style = AdminStatusUi.orderStatusStyle(event.getStatus());
            TextView statusText = new TextView(requireContext());
            statusText.setText(style.getLabel());
            statusText.setTextColor(getResources().getColor(index == 0 ? R.color.pp_text_primary : R.color.pp_text_secondary, null));
            statusText.setTextSize(14f);
            if (index == 0) statusText.setTypeface(statusText.getTypeface(), Typeface.BOLD);
            textColumn.addView(statusText);

            TextView timeText = new TextView(requireContext());
            timeText.setText(dateTimeFormat.format(event.getTimestamp()));
            timeText.setTextColor(getResources().getColor(R.color.pp_disabled, null));
            timeText.setTextSize(11f);
            timeText.setPadding(0, (int) (2 * density), 0, 0);
            textColumn.addView(timeText);

            row.addView(dotColumn);
            row.addView(textColumn);
            getBinding().llTimeline.addView(row);
        }
    }
}

File: app/src/main/java/com/example/purepawapp/ui/order/OrderDetailFragment.kt
package com.example.purepawapp.ui.order

import android.os.Bundle
import android.view.Gravity
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import android.widget.TextView
import androidx.lifecycle.lifecycleScope
import androidx.navigation.fragment.findNavController
import com.example.purepawapp.R
import com.example.purepawapp.data.model.Order
import com.example.purepawapp.data.model.OrderStatusEvent
import com.example.purepawapp.databinding.FragmentOrderDetailBinding
import com.example.purepawapp.databinding.ItemAdminOrderLineBinding
import com.example.purepawapp.di.ServiceLocator
import com.example.purepawapp.ui.admin.orderStatusStyle
import com.example.purepawapp.ui.common.BaseFragment
import com.example.purepawapp.util.toVndString
import com.example.purepawapp.util.toast
import kotlinx.coroutines.launch
import java.text.SimpleDateFormat
import java.util.Locale

class OrderDetailFragment : BaseFragment<FragmentOrderDetailBinding>(FragmentOrderDetailBinding::inflate) {

    private val orderId: String by lazy { arguments?.getString("orderId").orEmpty() }
    private val dateFormat = SimpleDateFormat("dd/MM/yyyy", Locale("vi", "VN"))
    private val dateTimeFormat = SimpleDateFormat("HH:mm, dd/MM/yyyy", Locale("vi", "VN"))

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)

        binding.btnBack.setOnClickListener {
            findNavController().popBackStack()
        }

        loadOrder()
    }

    override fun onResume() {
        super.onResume()
        loadOrder()
    }

    private fun loadOrder() {
        if (orderId.isBlank()) return
        viewLifecycleOwner.lifecycleScope.launch {
            ServiceLocator.orderRepository.getOrder(orderId)
                .onSuccess { bind(it) }
                .onFailure { toast(it.message ?: "Không thể tải đơn hàng") }
        }
    }

    private fun bind(order: Order) {
        val style = orderStatusStyle(order.status)
        binding.tvHeroTitle.text = style.label
        binding.tvHeroSubtitle.text = "Ngày đặt: ${dateFormat.format(order.createdAt)}"
        binding.tvOrderCode.text = "Mã đơn: ${order.orderCode.ifBlank { "#${order.id.take(8)}" }}"

        renderTimeline(order)

        binding.llOrderItems.removeAllViews()
        order.items.forEach { item ->
            val lineBinding = ItemAdminOrderLineBinding.inflate(LayoutInflater.from(requireContext()), binding.llOrderItems, false)
            lineBinding.tvLineName.text = "${item.productName} x${item.quantity}"
            lineBinding.tvLineSubtotal.text = item.subtotal.toVndString()
            binding.llOrderItems.addView(lineBinding.root)
        }

        binding.tvShippingName.text = "${order.shippingAddress.fullName} | ${order.shippingAddress.phone}"
        binding.tvShippingAddress.text = listOf(
            order.shippingAddress.street,
            order.shippingAddress.ward,
            order.shippingAddress.district,
            order.shippingAddress.city
        ).filter { it.isNotBlank() }.joinToString(", ")

        binding.tvPaymentMethod.text = when (order.paymentMethod.lowercase()) {
            "cod" -> "Thanh toán khi nhận hàng (COD)"
            "bank", "banking" -> "Chuyển khoản ngân hàng"
            else -> order.paymentMethod.ifBlank { "Thanh toán khi nhận hàng (COD)" }
        }

        binding.tvSubtotal.text = order.subtotal.toVndString()
        binding.tvShippingFee.text = order.shippingFee.toVndString()
        if (order.discountAmount > 0) {
            binding.rowDiscount.visibility = View.VISIBLE
            binding.tvDiscount.text = "-${order.discountAmount.toVndString()}"
        } else {
            binding.rowDiscount.visibility = View.GONE
        }
        binding.tvTotalAmount.text = order.totalAmount.toVndString()
    }

    private fun renderTimeline(order: Order) {
        binding.llTimeline.removeAllViews()
        val density = resources.displayMetrics.density
        val history = order.statusHistory.ifEmpty {
            listOf(OrderStatusEvent(status = order.status, timestamp = order.createdAt))
        }.sortedByDescending { it.timestamp }

        history.forEachIndexed { index, event ->
            val isLast = index == history.lastIndex
            val row = LinearLayout(requireContext()).apply {
                orientation = LinearLayout.HORIZONTAL
                layoutParams = LinearLayout.LayoutParams(
                    LinearLayout.LayoutParams.MATCH_PARENT,
                    LinearLayout.LayoutParams.WRAP_CONTENT
                )
            }

            val dotColumn = LinearLayout(requireContext()).apply {
                orientation = LinearLayout.VERTICAL
                gravity = Gravity.CENTER_HORIZONTAL
                layoutParams = LinearLayout.LayoutParams(
                    (12 * density).toInt(),
                    if (isLast) LinearLayout.LayoutParams.WRAP_CONTENT else LinearLayout.LayoutParams.MATCH_PARENT
                )
            }
            val dot = View(requireContext()).apply {
                background = resources.getDrawable(R.drawable.shape_circle, null)
                backgroundTintList = resources.getColorStateList(
                    if (index == 0) R.color.pp_primary else R.color.pp_timeline_inactive,
                    null
                )
                layoutParams = LinearLayout.LayoutParams((12 * density).toInt(), (12 * density).toInt())
            }
            dotColumn.addView(dot)
            if (!isLast) {
                val line = View(requireContext()).apply {
                    setBackgroundColor(resources.getColor(R.color.pp_timeline_inactive, null))
                    layoutParams = LinearLayout.LayoutParams((2 * density).toInt(), 0, 1f)
                }
                dotColumn.addView(line)
            }

            val textColumn = LinearLayout(requireContext()).apply {
                orientation = LinearLayout.VERTICAL
                layoutParams = LinearLayout.LayoutParams(
                    0,
                    LinearLayout.LayoutParams.WRAP_CONTENT,
                    1f
                ).apply {
                    marginStart = (16 * density).toInt()
                    if (!isLast) bottomMargin = (20 * density).toInt()
                }
            }
            val style = orderStatusStyle(event.status)
            textColumn.addView(TextView(requireContext()).apply {
                text = style.label
                setTextColor(resources.getColor(if (index == 0) R.color.pp_text_primary else R.color.pp_text_secondary, null))
                textSize = 14f
                if (index == 0) setTypeface(typeface, android.graphics.Typeface.BOLD)
            })
            textColumn.addView(TextView(requireContext()).apply {
                text = dateTimeFormat.format(event.timestamp)
                setTextColor(resources.getColor(R.color.pp_disabled, null))
                textSize = 11f
                setPadding(0, (2 * density).toInt(), 0, 0)
            })

            row.addView(dotColumn)
            row.addView(textColumn)
            binding.llTimeline.addView(row)
        }
    }
}
